### ISSUE TITLE
Failing tests for diffusion networks on Mac

### ISSUE THREAD

**Issue by Kucharssim** [stefanradev93/BayesFlow#739]
**Describe the bug**

Some tests are failing on Mac. 

**To Reproduce**
Minimal steps to reproduce the behavior:
1. On Mac Silicon
2. Checkout and setup bayesflow (tensorflow or torch as a backend)
3. run `pytest tests/test_networks/test_inference_networks.py::test_density_numerically`
4. The following tests fail:

```
tests/test_networks/test_inference_networks.py::test_density_numerically[batch_size=2-feature_size=2-conditions_size=None-generative_inference_network='diffusion_model_cosine_velocity'] FAILED 
tests/test_networks/test_inference_networks.py::test_density_numerically[batch_size=2-feature_size=2-conditions_size=None-generative_inference_network='diffusion_model_edm_potential'] FAILED  
```
**Expected behavior**

Tests should pass. To me, it is unclear whether there is an issue with the code or with the test.

Just a thought: Perhaps it wouldn't hurt to run automatic tests on GitHub with Mac. We would probably detect this issue sooner.

**Minimality**
- [x] I verify that my example is minimal, does not rely on third-party packages, and is most likely an issue in BayesFlow.

**Comment by stefanradev93** [comment]
@arrjon Can you debug this, as you are on a Mac?

@Kucharssim Yes, I will adapt the test suite moving forward.

**Comment by paul-buerkner** [comment]
I have run some quick tests myself and it appears to be just an issue with numerical accuracy to me. If I increase test tolerances from 1e-3 to 5e-3 or so, the tests are passing. I am currently checking which of the tolerances actually matter, but it seems to just be an issue of numerical accuracy, not necessarily something we can/have to address from the bayesflow side (except for increasing test tolerance).

**Comment by stefanradev93** [comment]
Any idea on why the continuous CoV is less accurate on Mac than on Windows/Linux?

**Comment by paul-buerkner** [comment]
Numerical differences in the ODE solvers, potentially caused by differences in linear algebra libraries (BLAS/LAPACK) or, more likely, by the use of MPS (https://developer.apple.com/metal/pytorch/), which is MAC specific and applies to torch and, I think, to tensorflow too. Apparently, these are known to sometimes cause such numerical tests to fail. I don't see anything we can do right now apart from increasing tolerance. But I am not an expert in ODE solvers and OS-specific numerical behavior, so don't take my opinion too seriously here. In any case, I will submit a small PR for this soon.

**Comment by paul-buerkner** [comment]
Okay, a little update: Even with 1e-02 I can only "fix" 3 out of the 4 tests. The Jacobian seems to be even more numerically instable then what is labled "output" in the test. I am right to assume these tests were only added with the autodiff changes recently?

**Comment by stefanradev93** [comment]
The tests have been around before the backend-agnostic autodiff utilities.

**Comment by arrjon** [comment]
> The tests have been around before the backend-agnostic autodiff utilities.

Indeed, and I remember that they were always unstable and that we already increased tolerances.

For me, it seems that torch is worse than tensorflow here...

**Comment by arrjon** [comment]
```
def test_density_numerically(generative_inference_network, random_samples, random_conditions):
    from bayesflow.utils import jacobian

    try:
        if keras.backend.backend() == "jax" and hasattr(generative_inference_network, "integrate_kwargs"):
            # jax backend does not support adaptive solvers for numerical jacobian computation yet
            if generative_inference_network.integrate_kwargs["steps"] == "adaptive":
                generative_inference_network.integrate_kwargs.update({"steps": 250})

        if generative_inference_network.integrate_kwargs["steps"] == "adaptive":
            generative_inference_network.integrate_kwargs.update({"min_steps": 200})
```

Setting the minimal number of steps of the adaptive solver higher seems to resolve the issue for me, as tests do not fail anymore. So it seems to be a problem of the adaptive solver in combination with the backend. I will have a more detailed look at it.
